### DISCUSSION TITLE
Automatic unique ID shown in form

### DISCUSSION THREAD

**Discussion by njokel** [Budibase/budibase#4357] [Help] [answered]
I need to have a disabled field containing autogenerated unique ID when the form is opened. The user needs this ID when they are filling the form. E.g. ID-1, Name: foo, Reason: bar, ... ID-1 is needed before the form is closed, as the user needs to attach a scanned document with this ID handwritten on.

Is there any way to do this?

**Comment by mjashanks** [comment, marked as answer]
Hi There.

This is possible, here's how I did it. (also, my app export is attached below).

1. I have a table called `Applications` - it has an `Auto ID` column. I.e. and auto incremented number. Your table might already have it, if you left the defaults checked on creation:
![image](https://user-images.githubusercontent.com/3524181/152812419-587f0e78-ce9b-41b7-bd73-ae0006e8c70e.png)

2. In the design section, I add a new screen and choose my table `Applications`, which auto-generates 3 screens for me.
3. In my `/applications` screen, I modify the `Create New` button, so that it
- First, `Save Row` to my application table
- Then, `Navigate To` the URL `/application/{{ Action 1.Saved Row._id}}`
4. In my application edit screen (`/applications/:id`), I added a paragraph component at the top to display my `Auto Id` field

So, instead of `Create New` navigating to the `/applications/new/row` screen, it creates a new blank application and edits that row.

The drawback of this is that you might end up with a bunch of blank applications, as people `Create New`, then don't complete it for whatever reason.

[Auto ID-export-1644245622501.txt](https://github.com/Budibase/budibase/files/8016329/Auto.ID-export-1644245622501.txt)

**Reply by njokel** [reply]
Would this work, if I would make a hidden flag value for the table that is set to true whenever user saves the form/row data and have a cronjob clearing all unnecessary rows that have not been saved? Alternatively, there should be a couple of mandatory fields. Saving is the only place where those are put to the table so it should be just an empty row. Cronjob to check empty values and clear those rows out?

**Reply by mjashanks** [reply]
Yes - that would work. There's one issue with this at the moment... you can only delete one record at a time. 

So, you would need to do a CRON that
1. Runs very regularly
2. Runs a query, with a limit of 1
3. Deletes that one record!

In future, we need to add a `For Each` automation step, that will allow you to delete every row that is returned.

**Comment by njokel** [comment]
Hi,

Thanks @mjashanks 👍 

I managed to get mine working a few minutes before your reply so did not have chance to reply here yet. Also,, I have not had time to test it thoroughly, but at least for 20-30 tries it worked perfectly.

**My steps:**

1. Data provider configured to new db table containing a single setup line for the counter value
2. Create a new button inside a repeater inside a data provider, data provider provides only 1 value
3. Form topmost row is disabled and contains the unique-id, default value set as {{ State.valueshown }}
4. Button contains actions
- Save row, updating table provided by the repeater, column with value {{ add repeater.counter 1 }}
- Update state valueshown by using {{ repeater.counter }}
- Update state Pressed to True
5. Button has condition - update setting to disabled when pressed = True
6. Save button deletes valueshown and sets pressed to False

Now the table containing the setup value has always the next value which is shown on form when button is pressed. I will probably try to generate the unique-id based on few values entered to the form and when generate button is pressed, it opens up the remaining values.

I wonder why it does not fetch the new value saved to the database when update state is used? I tried using refresh data provider as well, but no avail.

**Reply by mjashanks** [reply]
You may have an issue with that method when you have multiple concurrent users. If 2 people load the screen at the same time, they will be assigned the same ID.

**Reply by njokel** [reply]
Ok, so there is no way to update the value during runtime from database? I.e. action does not do a API call to database when the action is invoked? Good to know.

**Reply by mjashanks** [reply]
The action does update the database immediately.... but if two people load the same record at once, they will get the same "Count" value. Then, they will try to set their document ID to `count + 1` - so will have the same "ID"
